### PR TITLE
[Backport][Pipelines] Noop Writes when Reader completed

### PR DESCRIPTION
**Pull request by BrennanConroy**
Backport of https://github.com/dotnet/runtime/pull/41071

## Customer Impact
Customers shouldn't be able to see any differences as we want to keep existing code working without modifications.
This is fixing a Pipelines usage pattern that could result in large memory pressure or OOMs in the worst case.

## Regression
No.

## Testing
Yes? Tests were written and existing tests required no changes.

## Risk
Low, change was purposefully tailored to not break existing code

### CODE REVIEW

**Review by halter73 — APPROVED** [review]

**Comment by BrennanConroy** [comment]
@Anipik Could you take a look to make sure I added the change to the build correctly?

File: pkg/Microsoft.Private.PackageBaseline/packageIndex.json [hunk 1]
```diff
@@ -2227,12 +2227,14 @@
...
-        "4.5.3"
+        "4.5.3",
+        "4.5.4"
       ],
       "InboxOn": {},
       "AssemblyVersionInPackageVersion": {
         "4.0.0.0": "4.5.1",
-        "4.0.0.1": "4.5.2"
+        "4.0.0.1": "4.5.2",
+        "4.0.0.2": "4.5.4"
```
Comment (Anipik) [inline pkg/Microsoft.Private.PackageBaseline/packageIndex.json]: These changes should be reverted from 2.1

File: src/packages.builds [hunk 2]
```diff
@@ -24,6 +24,9 @@
       <AdditionalProperties>$(AdditionalProperties)</AdditionalProperties>
     </Project>
     <!-- add specific builds / pkgproj's here to include in servicing builds -->
+      <Project Include="$(MSBuildThisFileDirectory)System.IO.Pipelines\pkg\System.IO.Pipelines.pkgproj">
```
Comment (Anipik) [inline src/packages.builds]: nit: indentation

**Review by Anipik — CHANGES_REQUESTED** [review]
Change the packageVersion & assemblyVersion here https://github.com/dotnet/corefx/blob/release/2.1/src/System.IO.Pipelines/dir.props#L5-L6

File: pkg/Microsoft.Private.PackageBaseline/packageIndex.json (hunk 1 above)
Comment (BrennanConroy) [reply pkg/Microsoft.Private.PackageBaseline/packageIndex.json]: This specific line only? And then don't change the assemblyVersion in dir.props?

Comment (Anipik) [reply pkg/Microsoft.Private.PackageBaseline/packageIndex.json]: The entire change. The assemblyVersion also needs to be updated here.

Comment (BrennanConroy) [reply pkg/Microsoft.Private.PackageBaseline/packageIndex.json]: > These changes should be reverted from 2.1

> The assemblyVersion also needs to be updated here.

? I'm super lost

Comment (Anipik) [reply pkg/Microsoft.Private.PackageBaseline/packageIndex.json]: sry the packageIndex.json shouldnt be updated in release/2.1 branch.
The assemblyVersion needs to be updated in System.IO.Pipelines/dir.props

**Review by Anipik — APPROVED** [review]
packaging changes looks good

**Comment by BrennanConroy** [comment]
>error : PackageIndex from D:\a\1\s\pkg\Microsoft.Private.PackageBaseline\packageIndex.json is missing an assembly version entry for 4.0.0.2 for package System.IO.Pipelines.

Seems to imply we do need a change in packageIndex.json

**Comment by Anipik** [comment]
> Seems to imply we do need a change in packageIndex.json

yeah it  seems like it requires that

**Comment by Anipik** [comment]
This have been approved offline

**Pull request merged**